### ISSUE TITLE
Would we ever want a lazy flag added to the napari_get_reader hookspec?

### ISSUE THREAD

**Issue by sofroniewn** [sofroniewn/napari-io-test#3]
Thinking about this more and work in this repo - would we ever want a lazy flag added to the napari_get_reader hookspec?

**Comment by tlambert03** [comment]
help me understand what you're envisioning...  what would the flag do?  (and by flag do you mean another positional argument in the function signature?)

**Comment by sofroniewn** [comment]
flag would be a positional argument in the function signature

```python
def napari_get_reader(path: PathLike, lazy=False) -> Optional[ReaderFunction]:
```

that plugins could do what that want with - (we wouldn't have control) - but they could in theory decide to provide data back in a lazy or non-lazy fashion. Say I point at a directory of png, do you want lazy or non-lazy loading into a dask or numpy array respectively. We'd pass the value according to our `open_lazy` / shift drag & drop options.

I think we can see how things evolve before committing, but putting it out there

**Comment by tlambert03** [comment]
I'm not sure how well that loose of a contract would work.  Plugins can _already_ provide lazy data if they want to (by just returning a dask array) or non-lazy if they want to.  So it's unclear why they need a new argument in the hookspec, particularly if it's optional for them to honor it.  but let me know if I'm missing a nuance here

**Comment by sofroniewn** [comment]
I tend to agree with you, but to keep pushing - the argument is that this would then let users decide which one to use in a given circumstance without having to do any additional sorting / enabling / disabling, and it would let a plugin switch back and forth between either mode.

I guess the big question is are there circumstances when you'd want to be able to flexibly choose or can the plugin always choose for you based on the path? I'm thinking of times when the additional overhead of being lazy isn't worth it if it's not that big a deal to load it all in and then have everything else be much faster. Maybe it also depends how good caching is, which I guess the plugin has to be responsible for.

Or do we have `napari-io` and `napari-io-lazy` - that seems like less than optimal if we had to have two repos. If we could get that both from one repo that might be nicer - thinking about this comment https://github.com/napari/napari/pull/1023#issuecomment-599152375

**Comment by tlambert03** [comment]
I like push-back :)
I do think it would be nice to give some user control over laziness... but i wonder if that shouldn't be something we handle _for_ the plugins, instead of passing a `lazy` argument to the plugins that they may or many not abide by (as a side-note: I think it's asking too much to expect all plugins to offer both lazy and non-lazy implementations, so we almost _have_ to make it "optional" for plugins to pay attention to the `lazy` argument, and I don't love that).

However, we might already be well-equipped to handle this internally since `napari_get_reader` expects a _function_ to be returned, and not the data.  We could potentially wrap that function in `dask.delayed` and "lazify" plugins (kinda doing what dask_image.imread does, but wrapped around our plugin infrastructure).  The only tricky bit there is the handling of stacking behavior (but that's tricky in general when 3D images are laid out in 2D chunks in memory).

If you like that idea, me know if you want me to play with that or if you want to give it a shot

**Comment by sofroniewn** [comment]
> napari_get_reader expects a function to be returned, and not the data. We could potentially wrap that function in dask.delayed

Interesting idea, but I think we should hold back on that for now until we have done more work around our async stuff more broadly - see https://github.com/napari/napari/issues/845 for example.
